presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/CachingJdbcClient.java

```diff
@@ -346,31 +371,60 @@ public String quoted(RemoteTableName remoteTableName)
         return delegate.getTableScanRedirection(session, tableHandle);
     }
 
+    private Map<String, Object> getSessionProperties(ConnectorSession session)
+    {
+        return sessionProperties.stream()
+                .map(property -> Map.entry(property.getName(), getSessionProperty(session, property)))
+                .collect(toImmutableMap(Map.Entry::getKey, Map.Entry::getValue));
+    }
+
+    private static Object getSessionProperty(ConnectorSession session, PropertyMetadata property)
+    {
+        return firstNonNull(session.getProperty(property.getName(), property.getJavaType()), NULL_MARKER);
+    }
+
     private void invalidateSchemasCache()
     {
         schemaNamesCache.invalidateAll();
     }
 
-    private void invalidateTablesCaches()
+    private void invalidateTableCaches(SchemaTableName schemaTableName)
+    {
+        invalidateColumnsCache(schemaTableName);
+        invalidateCache(tableHandleCache, key -> key.tableName.equals(schemaTableName));
+        invalidateCache(tableNamesCache, key -> key.schemaName.equals(Optional.of(schemaTableName.getSchemaName())));
+    }
+
+    private void invalidateColumnsCache(SchemaTableName table)
     {
-        columnsCache.invalidateAll();
-        tableHandleCache.invalidateAll();
-        tableNamesCache.invalidateAll();
+        invalidateCache(columnsCache, key -> key.table.equals(table));
     }
 
-    private void invalidateColumnsCache(ConnectorSession session, SchemaTableName table)
+    @VisibleForTesting
+    CacheStats getColumnsCacheStats()
     {
-        columnsCache.invalidate(new ColumnsCacheKey(JdbcIdentity.from(session), table));
+        return columnsCache.stats();
+    }
+
+    private static <T, V> void invalidateCache(Cache<T, V> cache, Predicate<T> filterFunction)
```

Comment:
Key is not @Nullable. Please change `com.google.common.base.Predicate` -> `java.util.function.Predicate`